k8s/go/pkg/resolver/resolver.go

```diff
@@ -58,14 +58,41 @@ func RegisterFlags(flagset *flag.FlagSet) *Flags {
 	return &flags
 }
 
+// Option can be passed to NewResolver to configure the resolver.
+type Option struct {
```

Comment:
The `Option` pattern is pretty common in Go to parameterize a function with optional arguments.

By using your library within a custom resolver binary, it's possible to replace the heuristics built into the resolver library with custom logic without needing to fork the library.

```go
var (
	insecureRegistries = map[string]struct{}{
		"localhost:32000":      {},
		"othermachine:32000":       {},
	}
)

func main() {
	flagset := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
	flags := lib.RegisterFlags(flagset)
	flagset.Parse(os.Args[1:])

	resolver := lib.NewResolver(flags, lib.ParseTagOption(parseImageTag))
	resolved, err := resolver.Resolve()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(resolved)
}

func parseImageTag(t string, opts ...name.Option) (name.Tag, error) {
	got, err := name.NewTag(t, opts...)
	if err != nil {
		return got, err
	}

	regName := got.Context().Registry.Name()
	if _, isInsecure := insecureRegistries[regName]; isInsecure {
		var allOpts []name.Option
		allOpts = append(allOpts, opts...)
		allOpts = append(allOpts, name.Insecure)
		return name.NewTag(t, allOpts...)
	}
	return got, fmt.Errorf("unknown registry name %q", regName)
}
```

Reply:
I don't see how that is possible when Options struct rather than an interface. No one will be able to provide an alternative implementation of `apply()`

Reply:
The resolver library provides the only means of constructing an `Option`, like `ParseTagOption`.

This pattern is used by the cmp library, though it uses an interface with unexported methods rather than a struct. Either works. https://pkg.go.dev/github.com/google/go-cmp/cmp#Option

Reply:
I'm not a huge fan of this API design, but I guess we can change it later if necessary.